### DISCUSSION TITLE
Connection pool timeout

### DISCUSSION THREAD

**Discussion by MorenoMdz** [prisma/prisma#20537] [Q&A]
### Question

We started getting A LOT of pool timeout connection errors from Prisma since our last release yesterday. The release was basically some unrelated changes and upgrading the versions of the Node in our NestJs server from 16 to 20.

The errors look like this, 90% of the time from the same query:

```js
Invalid prisma.tableName.findMany() invocation:
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 17)
```

But they also show with a different limit

```js
Invalid prisma.tableName.findMany() invocation:
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 33)
```

Info:

 Our server runs on a Render.com service that has a (0.5) half core configuration, and can scale up to 15 instances. 
It connections to a Postgres pgBouncer that is set to 33 connections (15*2 + 1). Prisma is running using the Prisma Service package with default settings.

Questions:
* Why does the connection limit change from error to error? What causes it to go from 17 to 33?
* The DB server pool is set to 31 due to our max number of instances being 15. The db server has 2 cores, should we set the PG bouncer max connections to something else?
* Following the Prisma docs [here](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#optimizing-the-connection-pool) I think the next steps are tweaking the max connections number and the max timeout as well, are there any steps I could try before that?

### How to reproduce (optional)

It happens randomly over the day, we can't tie it to a specific behavior yet.

### Information about Prisma Schema, Client Queries and Environment (optional)

```js
// Add your schema.prisma and relevant query code
```

- OS: Ubuntu
- Database: PostgreSQL
- Node.js version: 20.4
- Prisma version: 4.16.2

**Comment by nurul3101** [comment]
Hello @MorenoMdz 👋 

Thank you for providing a detailed description of the issue you are facing. 🙏 

Can you check if the timeout errors go away if you use Node v16? It can help in confirming if this issue was indeed caused by the upgrade.

> Why does the connection limit change from error to error? What causes it to go from 17 to 33?

Based on my understanding, the number of active connections can change based on the number of incoming requests and the number of concurrent requests your application is handling at a time. If there's a sudden surge in incoming requests, your application will try to open more connections to the database to handle these requests, which can cause the connection limit to increase.

> The DB server pool is set to 31 due to our max number of instances being 15. The db server has 2 cores, should we set the PG bouncer max connections to something else?

If your server only has 2 cores, then having a connection limit that's too high could potentially lead to resource exhaustion on the server. Can you try enabling [metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) and check if any connections stay idle?

> Following the Prisma docs [here](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#optimizing-the-connection-pool) I think the next steps are tweaking the max connections number and the max timeout as well, are there any steps I could try before that?

The number of connections allowed by your PostgreSQL server (or pgBouncer) should be configured based on your application's needs and your server's resources. There isn't a one-size-fits-all number for this. So, I think tweaking max connections number and timeout should be a good next step.

**Comment by MorenoMdz** [comment]
> Based on my understanding, the connection limit can change based on the number of incoming requests and the number of concurrent requests your application is handling at a time. If there's a sudden surge in incoming requests, your application will try to open more connections to the database to handle these requests, which can cause the connection limit to increase.

That was my assumption as well, but doesn't that looks weird when the error says "limit", if the limit is dynamic it is not really a limit?

> If your server only has 2 cores, then having a connection limit that's too high could potentially lead to resource exhaustion on the server. Can you try enabling [metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) and check if any connections stay idle?

The DB server in this case right? Not the NestJs server as this one dynamically scales up and down. I am having some issues wrapping my head around the idea of 5 connections (2 cores * 2 +1) in the pool would be enough for for example a spike of 100 connections.

> The number of connections allowed by your PostgreSQL server (or pgBouncer) should be configured based on your application's needs and your server's resources. There isn't a one-size-fits-all number for this. So, I think tweaking max connections number and timeout should be a good next step.

Right now it has the defaults, which for the connections limit seems to be dynamic so I am not getting the reason to manually edit it, more info about it is welcome hehe

On the timeout I think I will try with 20s for now, is that reasonable?

**Comment by MorenoMdz** [comment]
Sorry forgot the big question. Yes we tested in Node 16 and it does not crash.

**Reply by janpio** [reply]
So with Node 16 everything works, but the exact same setup with Node 20 crashes? And that is reproducible?

**Reply by MorenoMdz** [reply]
Yes, might be able to scope it down to a single service running one of the endpoints that has this error, I will see if I can put some time to get this in a repo over the weekend

**Reply by MorenoMdz** [reply]
On the "crash" part, is it expected to have the server crashing with an unhandled exception due to a max connection limit?

**Reply by janpio** [reply]
That is up to you. Prisma just throws an error message that you can handle in your application. If you don't do that, depending on your framework and setup, that might crash the server, yes - but that is the same for any exception then, no matter if from Prisma or your own code.

**Comment by janpio** [comment]
Hey @MorenoMdz.

Let's start with the simple question:

> Why does the connection limit change from error to error? What causes it to go from 17 to 33?

The connection limit is either set via the `connection_limit` query parameter of the connection string, or when that one is not provided calculated based on the number of CPUs of the machine the app is running on: num_physical_cpus * 2 + 1. So you would get 17 on a server that reports 8 CPUs, and 32 on one with 16. Prisma does never dynamically change that value - so I would expect you got this error from two servers with different CPUs or configurations.

And now:

> Our server runs on a Render.com service that has a (0.5) half core configuration,

What does `(0.5) half core configuration` mean or imply? Does it refer to this?
<img width="951" alt="image" src="https://github.com/prisma/prisma/assets/183673/7216d2f3-1209-4f77-a8e5-94ef5dfce952">

>  and can scale up to 15 instances.

I assume this means that your app is configured to potentially use 15 instances of these Starter/0.5 core services if needed, so your app could potentially run 15 times?

> It connections to a Postgres pgBouncer that is set to 33 connections (15*2 + 1). 

Ok, so you have a PgBouncer that is configured to offer 33 connections to your apps maximum, correct?

> Prisma is running using the Prisma Service package with default settings.

What does that mean, I have not found anything googling for "Prisma Service package".

> The DB server pool is set to 31 due to our max number of instances being 15. The db server has 2 cores, should we set the PG bouncer max connections to something else?

So your database server has 2 cores and allows maximum 31 connections?
Are these cores busy when the error happens or is CPU and memory still available?


> On the timeout I think I will try with 20s for now, is that reasonable?

Yes, increasing the pool timeout should allow more queries to be successful at the cost of worse UX of your app - waiting for 20s in a request is not a great experience. But you get less of those errors hopefully until we figure out what is really going on.

--- 

Generally, a connection pool timeout error message means that on one server (instance), the Prisma Client connection pool could not give a new query a connection to be executed in the connection pool timeout of 10s. This happens when the pool was busy running other queries all that time, either because there were just too many queries, or the queries were too slow - or of course because the connection could not be opened in the necessary time.

Right now the numbers above do not make sense to me. 17 or 33 connections per server, 15 instances, would mean many more potential database connections (33*15) than what I understand your PgBouncer and database to be configured (~30). I will need responses from you to these questions before we can continue.

How much traffic does your app serve?
How many requests?
How many Prisma Client queries does that trigger?
Are any of these queries slow (>500ms)?

**Reply by MorenoMdz** [reply]
> The connection limit is either set via the connection_limit query parameter of the connection string, or when that one is not provided calculated based on the number of CPUs of the machine the app is running on: num_physical_cpus * 2 + 1. So you would get 17 on a server that reports 8 CPUs, and 32 on one with 16. Prisma does never dynamically change that value - so I would expect you got this error from two servers with different CPUs or configurations.

That was my assumption as well but how would that happen considering our services config is indeed the "0.5" ( which we internally mock as "half a CPU")? It does scale up to a max of 15 instances, which would not get us to 33 connections either.

> Ok, so you have a PgBouncer that is configured to offer 33 connections to your apps maximum, correct?

Yep that is correct.

> What does that mean, I have not found anything googling for "Prisma Service package".

I am sorry I meant to add the link to Nestjs Prisma
https://www.npmjs.com/package/nestjs-prisma

> How much traffic does your app server?
How many requests?
How many Prisma Client queries does that trigger?
Are any of these queries slow (>500ms)?

We are in between a DB system migration, we are porting from Firestore to PG, so a bunch of our logic is currently querying from both FS and PG and parsing the data as needed. So some requests are slow per se but the PG queries specifically are never that slow (more on that). Our PG server is hosted in Aiven.io inside the same cloud network (GCP us-east).

We do have some specific search endpoints that run queries that are slower than 500ms, and those all run fine, just some calls using `findMany` ended up hitting the timeout, so maybe it is related to that? 

> How many Prisma Client queries does that trigger?

Thats a good question, does Prisma split the `findMany` calls internally into multiple queries depending on the query clauses or not always?

>How many requests?

Around 50k requests per day with a peak of 90 per minute, so not a lot.

I am still confused how the server is deciding to scale up to 17 or 33 connections considering each instance can have a max of one shared core.

**Reply by janpio** [reply]
> > The connection limit is either set via the connection_limit query parameter of the connection string, or when that one is not provided calculated based on the number of CPUs of the machine the app is running on: num_physical_cpus * 2 + 1. So you would get 17 on a server that reports 8 CPUs, and 32 on one with 16. Prisma does never dynamically change that value - so I would expect you got this error from two servers with different CPUs or configurations.
> 
> That was my assumption as well but how would that happen considering our services config is indeed the "0.5" ( which we internally mock as "half a CPU")? It does scale up to a max of 15 instances, which would not get us to 33 connections either.

The number from the error message is _per instance_, not across all the instances. Prisma Client runs on each instance and opens a connection pool with x connections each time. You can see the connection pool size when you enable logging and debugging, and then start you app - it logs the number of connection it starts the connection pool for. If that is 17 or 33, the instance you are running on is reporting 8 or 16 CPUs to Prisma Client.

> Thats a good question, does Prisma split the `findMany` calls internally into multiple queries depending on the query clauses or not always?

Yes, but I was specifically asking for the Prisma Client queries. I assume 1 request to your app creates multiple calls to Prisma already - and that is a relevant number to know to be able to understand what "90 requests per minute" can lead to.

> We do have some specific search endpoints that run queries that are slower than 500ms, and those all run fine, just some calls using `findMany` ended up hitting the timeout, so maybe it is related to that?

The timeout would affect all queries being executed on a saturated instance, you are probably just mainly seeing this one because it is the most common query you are running a lot maybe?

---

Please take another look at my post as well, there were a few questions you did not answer.

In such situations it is essential to have a full picture of what is going on.

**Reply by MorenoMdz** [reply]
> What does (0.5) half core configuration mean or imply? Does it refer to this?
I assume this means that your app is configured to potentially use 15 instances of these Starter/0.5 core services if needed, so your app could potentially run 15 times?
Ok, so you have a PgBouncer that is configured to offer 33 connections to your apps maximum, correct?

Yes this service uses the starter service with 0.5 cores and can scale up to 15 instances, so up to 15 apps running connected to the same DB server using pgbouncer. The max number of instances we have over the past day was 6.

> What does that mean, I have not found anything googling for "Prisma Service package".

We are using this NestJs package 
https://www.npmjs.com/package/nestjs-prisma

> So your database server has 2 cores and allows maximum 31 connections?
Are these cores busy when the error happens or is CPU and memory still available?

It looks the DB server not got a heavy load, CPU over last day:

<img width="1314" alt="image" src="https://github.com/prisma/prisma/assets/29472480/2dac893b-4eef-4f1b-8170-93a63ba33930">

RAM over the last day:
<img width="1320" alt="image" src="https://github.com/prisma/prisma/assets/29472480/ab20b888-3d2b-48e6-9ee5-fbc4ed963151">

Our errors in the same period: 
<img width="611" alt="image" src="https://github.com/prisma/prisma/assets/29472480/d6e56a6b-3dd9-4e2c-b493-57989ad0eee2">

**Reply by janpio** [reply]
Ok.

So you need to configure your database and PgBouncer in a way that the number of allowed connections is lower than the sum of all the connection pool limits all your PrismaClient instances running in your app.

That could for example mean:
- Database: Maximum 100 connections
- PgBoucner: Maximum 160 connections
- Application (Prisma Client): Each instance 10 connections maximum, 15 instances maximum.

This way, if all 15 instances are turned on, and all use the full connection limit of the pool, it will still fit the limit of PgBouncer of 160 (probably 150 in use, with a few to spare - always good). And then PgBouncer will use the maximum 100 connections of the database probably to get the most value out of this.

And then you can see what is really happening: 
How many connections are used on the database?
How many connection are used on PgBouncer?
Are you still getting error messges by Prisma Client's connection pool?

Right now I think your setup is the wrong way around, with Prisma potentially using many more connections in total than PgBouncer and the database even offer. That should in theory lead to a different error than the connection pool limit ("out of connections"), but 🤷 That is super hard to debug what is going on.

Let me know how this turns out when you change the configuration.

---

> It looks the DB server not got a heavy load, CPU over last day:
> 
> <img alt="image" width="1314" src="https://user-images.githubusercontent.com/29472480/258261196-2dac893b-4eef-4f1b-8170-93a63ba33930.png">
> RAM over the last day: <img alt="image" width="1320" src="https://user-images.githubusercontent.com/29472480/258261276-ab20b888-3d2b-48e6-9ee5-fbc4ed963151.png">

What do the red and blue lines mean?
What do the numbers there mean?

**Reply by MorenoMdz** [reply]
>What do the red and blue lines mean?
What do the numbers there mean?

Resource % usage, the red one is the active instance and the blue ones are the read replica.

This might be slightly more helpful, data over the past month from DataDog from the PG metrics, nothing out of the ordinary, the higher usage zones are due to organic traffic we had, let me know if you notice anything different.
<img width="1411" alt="image" src="https://github.com/prisma/prisma/assets/29472480/a1ae4d10-7e8b-494f-ad68-d9b37fa36895">

One thing we did yesterday was upgrade our server to use Prisma 5.1, it seems to have fixed this issue so we can probably checkmark this one as complete and recommend using the latest version. There was a small hiccup in how the IN queries work in Prisma 5.1 but unrelated to this issue.

That said, there is still the problem of the number of connections reported by Prisma, I am going to do a few things to fully understand what's going on:
* Contact Render.com to figure out what the 0.5 CPU means and if our servers do scale up to 8 or 16 cores when scaling up and down.
* Figure out if it could be a weird dependency injection cycle issue from NestJs, as it is known for having dependency cycle problems, and I don't fully understand how the nestjs/prisma package instantiates and reuses the Prisma connection, maybe it could be a case where it is starting multiple DB instances and causing the client pool to read the number of connections incorrectly?
* I probably won't have time this weekend but maybe try to build a reproducible repo with the error happening in Node 20 but not in Node 20.
* Enable Prisma metrics in one of our staging/dev environments and run some stress tests.


For a bit more context, this was the query that was showing the error most of the time:

```ts
const data = await this.prisma.event.findMany({
          where: {
            userId,
            shift: {
              companyId,
            },
            status: {
              in: [
                status.ToDo,
                status.InProgress,
                status.OnBreak,
                status.Complete,
              ],
            },
          },
          orderBy: [{ shift: { startTime: 'asc' } }],
        })
```



On the number of connections topic...

>So you need to configure your database and PgBouncer in a way that the number of allowed connections is lower than the sum of all the connection pool limits all your PrismaClient instances running in your app.
That could for example mean:
Database: Maximum 100 connections
PgBoucner: Maximum 160 connections
Application (Prisma Client): Each instance 10 connections maximum, 15 instances maximum.

Good info, but then that goes against the recommended cores*2+1 formula in the pool config, setting up the pool to 160 connections sounds very intuitive IMHO but doesnt follow the recommendation, so you can see how this can get confusing 🙃 

Also, thank you very much for all the info and feedback so far 👊

**Reply by janpio** [reply]
> * Figure out if it could be a weird dependency injection cycle issue from NestJs, as it is known for having dependency cycle problems, and I don't fully understand how the nestjs/prisma package instantiates and reuses the Prisma connection, maybe it could be a case where it is starting multiple DB instances and causing the client pool to read the number of connections incorrectly?

No, what the error message is reporting is the connection pool _limit_. This is set when Prisma Client is started and does not change, and does not depend on how many connections are open or how often Prisma is started. Each number you see in this error message is from exactly one Prisma Client instance running in one app. As I described before, each instance probably runs its own Prisma Client instance with its own connection pool.

> Good info, but then that goes against the recommended cores*2+1 formula in the pool config, setting up the pool to 160 connections sounds very intuitive IMHO but doesnt follow the recommendation, so you can see how this can get confusing 🙃

The formula is not the recommendation to be used with production traffic, but a default that Prisma can calculate and use. It is impossible for Prisma to know what your traffic and your application requires, and what your server and database can deliver - so this is always only a starting point. (If that is not clear enough in some documentation, please let me know and we can get that fixed!)

> Also, thank you very much for all the info and feedback so far 👊

You're welcome!

**Comment by MorenoMdz** [comment]
One more question, for this case where we have a shared core, if we wanted to test a connection limit, should I try with 2? As 0,5 * 2 + 1 = 2? Would that be a max number of 2 connections per instance of our server?

**Reply by janpio** [reply]
I don't know what "a shared core" means really, and how that influences the performance. I would watch the actual CPU usage to define if it is too much or not.

Generally the default formula is just that - a default. You can to define the number of connections in the connection pool (the limit) based on the CPU usage and the actual needs of your application. If it wants or needs to run 50 queries at the same time, 2 available connections mean that 48 always have to wait. If your queries are not the fastest, the queue easily can take 10s to be emptied - and then one of those queries would get that error message again.

**Comment by MorenoMdz** [comment]
We are not seeing this error now when running with Node 20 and Prisma 5.1, I will mark this one as a potential solution for now but ill try to investigate more the issue.

**Reply by MorenoMdz** [reply]
I decided to re-open this one, as this problem seems to be related to the way Nest/Prisma detect the number of available CPU cores.

**Reply by lroal** [reply]
Yes, I suspect that too.

**Comment by MorenoMdz** [comment]
@janpio got and update from Render:

> Hi Moreno, 
> 
> We have a multi-tenant model, so the underlying machines do have more resources available than what we allocate for each instance type on Render. So if you were to run cat /proc/cpuinfo it would show the number of CPU cores as 8 even though for the Starter instance, this is limited to 0.5 CPU cores. I'm not sure exactly how Prisma reports the number of CPU cores, but if it's using this system-level value then that could potentially be throwing it off. 
 >
> > Just to confirm, 0.5 CPU means we have access to up to 50% of a shared CPU core per instance of our service
>
> Your understanding here is correct, however, when scaling, each instance would separately have access a half CPU core, so with 10 instances, that would be 5 CPU cores of total resources available, each limited to 0.5 CPU cores at the service level.
 
Does that mean Prisma might be reading the number of available CPUs incorrectly?

**Reply by janpio** [reply]
That is one way we could state it, or we would say that Render is reporting it incorrectly 😆 :trollface: 

Fact is, the default connection pool size that Prisma is using is based on incorrect information. We'll have to investigate that together with Render. Thanks for sharing.

**Reply by MorenoMdz** [reply]
Thank you for checking that, let me know if there's anything I can do to help.

**Reply by janpio** [reply]
Did you also get an explanation why the server sometimes reports 8 cores (connection limit 17) and sometimes 16 (connection limit 33)?

Generally, the "high" connection limit for the pool size did not actually cause you any problems. If Prisma had received and used the `0.5` value, that would have actually been worse - as a limit of 2 or 3 would have lead to pool exhaustion with the error message you reported much faster.

**Reply by MorenoMdz** [reply]
Just got it:

> Hi Moreno, 
> We have a few different sizes of nodes that we schedule user services on. Typically if workload is very large, then it would only schedule on bigger instance types that it would fit. 
If it's a smaller workload, then it would be scheduled on whatever node has space. So we don't really have set rules to determine if it will be scheduled on a machine with 8 CPU vs 16 CPU. A command like echo "$(($(cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us) / $(cat /sys/fs/cgroup/cpu/cpu.cfs_period_us)))" may work better for determining the CPU limit for the given service, but unfortunately this doesn't work for the Starter instance with 0.5 CPU as I believe it returns 0. It should work for the other instance types with a whole number as the CPU limit.
 
 @janpio

**Comment by alandotcom** [comment]
Also seeing similar issues. Not exclusive to the latest version of Prisma (v5.1.1), but also the last v4 release. We've been testing on Node v20 and recently downgraded to Node v18 in case there may have been stability issues there.

We're also running on Render, although across different environments (one with more CPU and memory, and one with less) and that doesn't seem to make a difference.

Steps we've taken so far:
- Setup pgBouncer, which maintains a fairly large connection pool to the db. Not seeing any issues there
- Set higher connection limits for services where we expect larger amounts of traffic

The thing is, we're seeing these connection pool timeouts happening with fairly low traffic. It generally gets triggered when there's a burst of traffic, but nothing that should overwhelm the number of connections set in the connection limit (e.g., we set the limit at 30 and get a burst of 30 requests).

Maybe there's an issue with how the connection pooler recycles connections and makes them available. I do see there were issues with this in the past.

**Reply by alandotcom** [reply]
One thing I noticed, I see connections constantly being opened and closed to pgBouncer,

```
closing because: client close request (age=851s)
```

**Reply by alandotcom** [reply]
We've seen an order of magnitude improvement when executing queries using `pg.Pool` directly, vs. using Prisma. Part of it is due to Prisma not using JOINs, but the worst p99 tail latencies were all due to connection pooling issues. Since we're seeing a major improvement now, I'm confident this isn't a database issue.

**Reply by MorenoMdz** [reply]
@alandotcom do you mind sharing some of those learnings? Perhaps a few examples of using pg.Pool directly?

**Reply by alandotcom** [reply]
We really just pulled in, https://node-postgres.com/, and used `Pool` to create a connection pool of ~20-50 connections (we're still using pgbouncer). We saw the issues primarily in places where we would execute maybe 4-5 queries all at the same time (e.g., the front-end would make a few API requests, each one doing its own db lookup). That seemed to cause issues pretty frequently with Prisma. We're no longer seeing those issues.

Possibly related, each of these queries would also do a few roundtrips to the DB, https://github.com/prisma/prisma/discussions/12715
So it would be around 4 API calls * 3-4 queries per call, so around 16 total queries. By just executing raw sql queries, we reduced this to just 1 db query per API request

**Reply by alandotcom** [reply]
I can't really measure connection pooling performance differences. We're using Datadog, but it's missing connection pooler tracing information, https://github.com/DataDog/dd-trace-js/issues/1923

**Reply by MorenoMdz** [reply]
We have a very similar stack, we also use DD, and we have it connected to our DB provider (Aiven) instead of Prisma rn, but I am not super sold the metrics we have there are super reliable yet.

**Reply by alandotcom** [reply]
If you wanted to give this a try, we're using [Kysely](https://kysely.dev) and https://github.com/valtyr/prisma-kysely to generate the types from the prisma schema. I've found it pretty easy to keep all the interfaces the same, so only the code in the database layer changes. We've basically replaced every select query in an API path with this.

**Reply by MorenoMdz** [reply]
> We really just pulled in, https://node-postgres.com/, and used `Pool` to create a connection pool of ~20-50 connections (we're still using pgbouncer). We saw the issues primarily in places where we would execute maybe 4-5 queries all at the same time (e.g., the front-end would make a few API requests, each one doing its own db lookup). That seemed to cause issues pretty frequently with Prisma. We're no longer seeing those issues.
> 
> Possibly related, each of these queries would also do a few roundtrips to the DB, #12715 So it would be around 4 API calls * 3-4 queries per call, so around 16 total queries. By just executing raw sql queries, we reduced this to just 1 db query per API request

I will try to run some stress test, we have two similar search endpoints accessing the same data, one using Prisma and one using PrismaRaw, I might be able to get some of those to reach the connection limit.

**Reply by janpio** [reply]
Please do open an issue if you have anything reproducible here that we can run ourselves and look at to see what is going on under the hood. We sometimes get reports like this, but never anything we can actually poke at and work with - so any input would be useful. Thanks.

**Reply by MorenoMdz** [reply]
> Please do open an issue if you have anything reproducible here that we can run ourselves and look at to see what is going on under the hood. We sometimes get reports like this, but never anything we can actually poke at and work with - so any input would be useful. Thanks.

Isnt the CPU count mismatch isn't a clue of what could be wrong? Not saying it is a Prisma or Render issue, but there is clearly an issue when we get a burst of requests that Prisma splits into multiple db calls.

**Reply by janpio** [reply]
No, the CPU count is just used to set the default connection pool size. You could also manually set it to any of these values and it should work. If you do so, and have the same problem, and can reproduce it, we can look into that. But we need to be able to see the problem to understand it.

**Reply by MorenoMdz** [reply]
@janpio going back to this, it seems the issue is the `pgbouncer` connection. We have tuned both our Nest + Prisma and our DB server connections, but we still do have high latency queries, even very simple queries, timing out and causing the whole pool to bubble up the number of connections until the DB gets unresponsive on that call and the Prisma pool starts timing out connections. I think it is related to the Pgbouncer change [here](https://github.com/pgbouncer/pgbouncer/pull/845), if so, it would be great to have Prisma supporting it as well so we can try connecting without the transaction mode

**Reply by janpio** [reply]
If you want to test Prisma, without `pgbouncer=true`, against that brand new version of PgBouncer, please let us know how it goes.

**Reply by MorenoMdz** [reply]
> If you want to test Prisma, without `pgbouncer=true`, against that brand new version of PgBouncer, please let us know how it goes.

Locally it seems to work fine, I have contacted Aiven and they are going to test and upgrade their servers with the new version and I will test it more by them and report here.

Did Prisma 5.4.1 change anything in relation to this?

**Reply by janpio** [reply]
Not that I am aware of, no.

**Comment by lroal** [comment]
I have the same problem when running in a github action. Locally, it works just fine. There are 10 connections in the pool.
Here is the code:
https://github.com/alfateam/orm-benchmarks/blob/pools/src/prisma/benchmark-pg.ts
```bash
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 10)
    at _n.handleRequestError (/__w/orm-benchmarks/orm-benchmarks/node_modules/@prisma/client/runtime/library.js:122:6927)
    at _n.handleAndLogRequestError (/__w/orm-benchmarks/orm-benchmarks/node_modules/@prisma/client/runtime/library.js:122:6235)
    at _n.request (/__w/orm-benchmarks/orm-benchmarks/node_modules/@prisma/client/runtime/library.js:122:5919)
    at async l (/__w/orm-benchmarks/orm-benchmarks/node_modules/@prisma/client/runtime/library.js:131:9116)
    at async Promise.all (index 18)
    at getRowsWithRelations (/__w/orm-benchmarks/orm-benchmarks/src/prisma/benchmark-pg.ts:[57](https://github.com/alfateam/orm-benchmarks/actions/runs/9796288292/job/27050384191#step:6:58):5)
    at benchmark (/__w/orm-benchmarks/orm-benchmarks/src/prisma/benchmark-pg.ts:28:5) {
  code: 'P2024',
  clientVersion: '5.16.1',
  meta: { modelName: 'Order', connection_limit: 10, timeout: 10 }
```bash

**Comment by lroal** [comment]
I tried with pool size 2 in github action - and it worked.  https://github.com/alfateam/orm-benchmarks/actions/runs/9804768733/job/27073137829
I think the number of cores in the github action is two - so it seems to be a problem detecting number of cores.

**Comment by lroal** [comment]
Pool size 3 also worked fine in github action.
But size 4 did not.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
